src/components/DataDimension/Calculation/CalculationModal.js

```diff
@@ -0,0 +1,417 @@
+import { useAlert, useDataMutation, useDataQuery } from '@dhis2/app-runtime'
+import {
+    Button,
+    Modal,
+    ModalTitle,
+    ModalContent,
+    ModalActions,
+    ButtonStrip,
+    InputField,
+} from '@dhis2/ui'
+import cx from 'classnames'
+import PropTypes from 'prop-types'
+import React, { useEffect, useState } from 'react'
+import { validateExpressionMutation } from '../../../api/expression.js'
+import i18n from '../../../locales/index.js'
+import {
+    parseExpressionToArray,
+    parseArrayToExpression,
+    validateExpression,
+    EXPRESSION_TYPE_DATA,
+    EXPRESSION_TYPE_NUMBER,
+    INVALID_EXPRESSION,
+    VALID_EXPRESSION,
+} from '../../../modules/expressions.js'
+import { OfflineTooltip as Tooltip } from '../../OfflineTooltip.js'
+import DataElementSelector from './DataElementSelector.js'
+import DndContext, { OPTIONS_PANEL } from './DndContext.js'
+import FormulaField, {
+    LAST_DROPZONE_ID,
+    FORMULA_BOX_ID,
+} from './FormulaField.js'
+import MathOperatorSelector from './MathOperatorSelector.js'
+import styles from './styles/CalculationModal.style.js'
+
+const FIRST_POSITION = 0
+const LAST_POSITION = -1
+
+const EMPTY_EXPRESSION_ARRAY = []
+
+const CalculationModal = ({
+    calculation,
+    onSave,
+    onClose,
+    onDelete,
+    displayNameProp,
+}) => {
+    const query = {
+        dataElements: {
+            resource: 'dataElements',
+            params: ({ ids = [] }) => ({
+                fields: `id,${displayNameProp}~rename(name)`,
+                filter: `id:in:[${ids.join(',')}]`,
+                paging: false,
+            }),
+        },
+        dataElementOperands: {
+            resource: 'dataElementOperands',
+            params: ({ ids = [] }) => ({
+                fields: `id,${displayNameProp}~rename(name)`,
+                filter: `id:in:[${ids.join(',')}]`,
+                paging: false,
+            }),
+        },
+    }
+
+    const { data, refetch } = useDataQuery(query, {
+        lazy: true,
+    })
+
+    useEffect(() => {
+        const expressionIds = calculation.expression
```

Comment:
Suggesting moving this regex to modules/expressions to keep regex's related to expressions in one place, and to make this testable. For instance add a new function `getExpressionDimensionIds` or similar.

Reply:
done 👍